File: src/tests/bench/python/bench.py
# python/bench.py
#
# Python benchmark workloads, same algorithms as liphia/*.lph.
# Usage: python bench.py <workload>
# Prints one checksum line; the runner times the whole process externally.

import sys


def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)


def nested_sum(n):
    total = 0
    i = 0
    while i < n:
        j = 0
        while j < n:
            total = total + i * j
            j = j + 1
        i = i + 1
    return total


def mandelbrot(size, max_iter):
    inside = 0
    y = 0
    while y < size:
        ci = float(y) * 2.0 / float(size) - 1.0
        x = 0
        while x < size:
            cr = float(x) * 3.0 / float(size) - 2.0
            zr = 0.0
            zi = 0.0
            k = 0
            while k < max_iter and zr * zr + zi * zi <= 4.0:
                t = zr * zr - zi * zi + cr
                zi = 2.0 * zr * zi + ci
                zr = t
                k = k + 1
            if k == max_iter:
                inside = inside + 1
            x = x + 1
        y = y + 1
    return inside


def count_primes(limit):
    is_prime = []
    i = 0
    while i < limit:
        is_prime.append(True)
        i = i + 1
    is_prime[0] = False
    is_prime[1] = False
    p = 2
    while p * p < limit:
        if is_prime[p]:
            m = p * p
            while m < limit:
                is_prime[m] = False
                m = m + p
        p = p + 1
    count = 0
    i = 0
    while i < limit:
        if is_prime[i]:
            count = count + 1
        i = i + 1
    return count


def bubble_sort_checksum(n):
    items = []
    i = 0
    while i < n:
        items.append(n - i)
        i = i + 1
    end = n - 1
    while end > 0:
        j = 0
        while j < end:
            if items[j] > items[j + 1]:
                t = items[j]
                items[j] = items[j + 1]
                items[j + 1] = t
            j = j + 1
        end = end - 1
    check = 0
    i = 0
    while i < n:
        check = check + items[i] * (i + 1)
        i = i + 1
    return check


def word_count(keys, rounds):
    counts = {}
    r = 0
    while r < rounds:
        k = 0
        while k < keys:
            word = "w" + str(k)
            if word in counts:
                counts[word] = counts[word] + 1
            else:
                counts[word] = 1
            k = k + 1
        r = r + 1
    total = 0
    for name in counts:
        total = total + counts[name] * len(name)
    return total


WORKLOADS = {
    "startup": lambda: "startup",
    "fib": lambda: "fib %d" % fib(32),
    "loops": lambda: "loops %d" % nested_sum(3000),
    "mandelbrot": lambda: "mandelbrot %d" % mandelbrot(300, 100),
    "sieve": lambda: "sieve %d" % count_primes(2000000),
    "sort": lambda: "sort %d" % bubble_sort_checksum(2500),
    "map": lambda: "map %d" % word_count(1000, 200),
}

print(WORKLOADS[sys.argv[1]]())

File: src/tests/bench/node/bench.js
// node/bench.js
//
// Node.js benchmark workloads, same algorithms as liphia/*.lph.
// Usage: node bench.js <workload>
// Prints one checksum line; the runner times the whole process externally.
// Integer results stay below 2^53, so plain numbers are exact.

function fib(n) {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

function nestedSum(n) {
    let total = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            total = total + i * j;
        }
    }
    return total;
}

function mandelbrot(size, maxIter) {
    let inside = 0;
    for (let y = 0; y < size; y++) {
        const ci = y * 2.0 / size - 1.0;
        for (let x = 0; x < size; x++) {
            const cr = x * 3.0 / size - 2.0;
            let zr = 0.0;
            let zi = 0.0;
            let k = 0;
            while (k < maxIter && zr * zr + zi * zi <= 4.0) {
                const t = zr * zr - zi * zi + cr;
                zi = 2.0 * zr * zi + ci;
                zr = t;
                k++;
            }
            if (k === maxIter) inside++;
        }
    }
    return inside;
}

function countPrimes(limit) {
    const isPrime = [];
    for (let i = 0; i < limit; i++) isPrime.push(true);
    isPrime[0] = false;
    isPrime[1] = false;
    for (let p = 2; p * p < limit; p++) {
        if (isPrime[p]) {
            for (let m = p * p; m < limit; m += p) isPrime[m] = false;
        }
    }
    let count = 0;
    for (let i = 0; i < limit; i++) if (isPrime[i]) count++;
    return count;
}

function bubbleSortChecksum(n) {
    const items = [];
    for (let i = 0; i < n; i++) items.push(n - i);
    for (let end = n - 1; end > 0; end--) {
        for (let j = 0; j < end; j++) {
            if (items[j] > items[j + 1]) {
                const t = items[j];
                items[j] = items[j + 1];
                items[j + 1] = t;
            }
        }
    }
    let check = 0;
    for (let i = 0; i < n; i++) check = check + items[i] * (i + 1);
    return check;
}

function wordCount(keys, rounds) {
    const counts = new Map();
    for (let r = 0; r < rounds; r++) {
        for (let k = 0; k < keys; k++) {
            const word = "w" + k;
            counts.set(word, (counts.get(word) || 0) + 1);
        }
    }
    let total = 0;
    for (const [name, count] of counts) total = total + count * name.length;
    return total;
}

const workloads = {
    startup: () => "startup",
    fib: () => `fib ${fib(32)}`,
    loops: () => `loops ${nestedSum(3000)}`,
    mandelbrot: () => `mandelbrot ${mandelbrot(300, 100)}`,
    sieve: () => `sieve ${countPrimes(2000000)}`,
    sort: () => `sort ${bubbleSortChecksum(2500)}`,
    map: () => `map ${wordCount(1000, 200)}`,
};

console.log(workloads[process.argv[2]]());
